monitor GPU ressources

AlexandreKempf opened the pull request:
# New feature: monitoring for harware metrics

For a first discussion on this PR content, you can look at [here](https://github.com/iterative/dvclive/pull/773). It was a outdated version of this PR that only contains CPU, ram and disk metrics.

### Monitoring hardware

In this PR we add the possibility for the user to monitor the GPU, CPU, RAM, and disk during one experiment.
The GPU metrics are only collected if at least one GPU is detected.

To use this feature you can use it with a simple argument:
```python
from dvclive import Live

with Live(monitor_system=True) as live:
    ...
```

If you want to use advance features you can specify each parameter this way:
```python
from dvclive import Live
from dvclive.monitor_system import SystemMonitor

with Live() as live:
    live.system_monitor = SystemMonitor(interval = 0.1, num_samples=15, directories_to_monitor={"data": "/path/to/data/directory", "home": "/home"}))
```

If you allow the monitoring of your system, if will track the following:
 - `system/cpu/count` -> number of CPU cores
 - `system/cpu/usage (%)` ->  the average usage of the CPUs. 
 - `system/cpu/parallelization (%)` -> How many CPU cores use more than 20% of their possibilities? It is useful when you're looking to parallelize your code to train your model or process your data faster. 
  - `system/ram/usage (%)` -> percentage of the RAM used. Useful to increase batch size or data processed at the same time in the RAM.
  - `system/ram/usage (GB)` -> RAM used. Useful to increase batch size or data processed at the same time. 
  - `system/ram/total (GB)` -> Total RAM in your system
  - `system/disk/usage (%)` -> Amount of disk used by the partition that contain the given path, in %. By default uses "/". You can specify the paths to the partition you want to monitor. For instance, the code example above monitors `/data` and `/home`. Data and code often live in very different paths/volumes, so it is useful for the user to be able to specify its own path. 
  - `system/disk/usage (GB)` ->  Amount of disk used at a given path.
  - `system/disk/total (GB)`  -> Total disk storage at a given path.
  - `system/gpu/count` -> Number of GPUs detected.
  - `system/gpu/usage (%)` -> Usage of each GPU in %.
  - `system/vram/usage (%)` -> Usage of each GPU virtual memory in %.
  - `system/vram/usage (GB)` -> Usage of each GPU virtual memory in GB.
  - `system/vram/total (GB)` -> total amount of GPU virtual memory in GB, for each GPU.

Note that as several paths can be specified, the full metric name is `system/disk/usage (%)/<user defined name>`. For instance it would be `system/disk/usage (%)/data` for the `/path/to/data/disk` and `system/disk/usage (%)/home` for `/home`.

Note that as several GPUs can be detected, the full metric name for GPU metrics (except `count`) is suffix with `/<idx>` that indicate the index of the GPU. Example: `system/gpu/usage (%)/0`

All the values that can change during an experiment can be saved as plots. Timestamps are automatically recorded with the metrics. Other metrics (that don't change) such as GPU count, GPU vram total, CPU count, RAM total and disk total are saved as metrics but cannot be saved as plots. 

I decided to split the usage in % and GB. First, because it is more consistent with the other loggers out there. Second, both are extremely relevant based on which cloud instance you run your experiment. If you only run your experiment on the same hardware, the distinction is not really interesting.

### Files generated
The metrics about the CPU are stored with the `log_metric` function. It means that the `.tsv` files are stored in the `dvclive/plots` folder. A specific folder, `system`, contains all the metrics about the CPU to distinguish them from the user-defined metrics. The metrics are also saved in the `dvclive/metrics.json` file. 

### Plot display
Here is what VScode extension looks like:
![image](https://github.com/iterative/dvclive/assets/14785566/1295eca0-b533-49a2-ac51-c91f662ba3c5)

Here is what Studio looks like:
https://studio.iterative.ai/user/AlexandreKempf/projects/image_classification-imzssxc5ew

Note that studio live update is a little buggy, but it is fixed in [this PR](https://github.com/iterative/dvclive/pull/788)

Note that we're calling `nvmlShutdown` and `nvmlInit` at each fetch of the GPU metrics like in [labML code](https://github.com/labmlai/labml/blob/6dc2a2ad3bd3a46c5dab9be80f3030c5ac8cb133/client/labml/internal/computer/monitor/scanner.py#L141)
---


- [x] ❗ I have followed the [Contributing to DVCLive](https://github.com/iterative/dvclive/blob/main/CONTRIBUTING.rst) guide.
- [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here: https://github.com/iterative/dvc.org/pull/5138

dberenbaum commented on pyproject.toml:
```diff
@@ -39,6 +39,7 @@ dependencies = [
     "ruamel.yaml",
     "scmrepo",
     "psutil",
+    "py3nvml"
```
It looks like this is quite old an unmaintained. Any reason not to use `pynvml` from https://pypi.org/project/nvidia-ml-py/? Also, we need to decide if we always install this or make it optional.

dberenbaum commented on pyproject.toml:
```diff
@@ -38,7 +38,8 @@ dependencies = [
   "scmrepo",
-  "psutil"
+  "psutil",
+  "py3nvml"
```
Reminder to switch to pynvml

dberenbaum commented on src/dvclive/monitor_system.py:
```diff
@@ -0,0 +1,265 @@
+            / num_cpus,
+        }
+
+    def _get_disk_info(self) -> Dict[str, Union[float, int]]:
```
If this is causing additional complications, I think we could drop it for now

AlexandreKempf replied:
What complication do you see? :)

dberenbaum replied:
I know there was discussion in the other PR like how to handle partitions. No problem if it's no longer an issue.

dberenbaum commented on src/dvclive/monitor_system.py:
```diff
@@ -0,0 +1,265 @@
+            result.update(disk_metrics)
+        return result
+
+    def __get_gpu_info(self) -> Dict[str, Union[float, int]]:
```
Why the double underscore in `__get_gpu_info`?

dberenbaum commented on pyproject.toml:
```diff
@@ -37,7 +37,9 @@ dependencies = [
   "ruamel.yaml",
-  "scmrepo"
+  "scmrepo",
+  "psutil",
+  "pynvml"
```
Should we include by default or make them optional?

shcheklein replied:
Yep, my 2cs. I would prefer "system" to be enabled by default.  Agreed to check on how big / complicated those deps are.

AlexandreKempf replied:
Both, when installed together, are quite light: ~500KB. 

@shcheklein Then should I change the `monitor_system` default to `True` in the `Live` object?
Any idea on that topic @dberenbaum?

My 0.02$ is that it would be nice to have it by default, especially for new users. But for existing users, it can be annoying to have new plots all over the studio dashboard I spend days customizing. What should we prioritize?

shcheklein replied:
> Then should I change the monitor_system default to True in the Live object?

yep, I would go for it. I think we should be optimizing for new users still. Not a strong opinion though.

dberenbaum replied:
I would vote to keep it off and consider making it a default in 4.0 for the reasons @AlexandreKempf mentioned (annoyance to existing users)

shcheklein replied:
> off and consider making it a default in 4.0

sounds good. let's make a ticket for this? :) and do it soon I hope. (I just feel it's a really good and appealing feature - people will stick better with the extension, Studio, etc).

AlexandreKempf replied:
Let's wait for 4.0. Do we have a roadmap to 4.0 by any chance, so I can add that ticket?
Also this ticket should mention that `psutil` and `pynvml` should not be optional anymore.

dberenbaum replied:
If we make the deps optional, do we need to change the PR a bit to make sure we fail gracefully if they are not installed? It looks like loading `dvclive.live` now will try to import those deps.

AlexandreKempf replied:
Ok, I put them by default as they are quite light. Also, having them installed by default should simplify the documentation.

dberenbaum commented on src/dvclive/monitor_system.py:
```diff
@@ -0,0 +1,265 @@
+        ).start()
+
+    def _monitoring_loop(self):
+        while not self._shutdown_event.is_set():
```
Can we put the rest of this logic into its own method? That way, it should be easier to test the underlying logic without having to create a thread that has to be shut down.

AlexandreKempf replied:
I'm not quite sure I understand what you mean here :) 
There is lots of logic with the _shutdown_event inside the following lines, how should I handle them?

dberenbaum replied:
Okay, sorry, didn't read/think through it carefully. If the tests are all stable now, no concerns from me.

shcheklein reviewed (approved):
Seems good to me. Thanks @AlexandreKempf !

dberenbaum commented:
Hey @AlexandreKempf, take a look at my remaining questions/suggestions, but if @shcheklein is satisfied from a technical perspective, feel free to merge when you feel it's ready.

shcheklein commented:
> but if @shcheklein is satisfied from a technical perspective, feel free to merge when you feel it's ready.

I've approved it already!

AlexandreKempf commented:
I added the doc [here](https://github.com/iterative/dvc.org/pull/5138) fyi.

AlexandreKempf commented:
Added changes based on recommendations in the [documentation PR](https://github.com/iterative/dvc.org/pull/5138)

- install `psutil` and `pynvml` by default
- removed `plot` argument since we always want these metrics plotted
- add method `monitor_system` to `Live` instead of a property|

dberenbaum commented on src/dvclive/live.py:
```diff
@@ -370,6 +380,43 @@ def step(self, value: int) -> None:
+                sampling interval small, the maximum value allowed is 0.1 seconds.
+                Default to 0.05.
+            num_samples (int): the number of samples to collect before the aggregation.
+                The value should be between 1 and 30 samples. Default to 20.
```
Why between 1 and 30?

AlexandreKempf replied:
Following W&B behavior on this one.

dberenbaum reviewed (approved):
I left a couple questions, but they aren't blockers, so feel free to merge when ready.

dberenbaum commented on src/dvclive/live.py:
```diff
@@ -370,6 +380,43 @@ def step(self, value: int) -> None:
+        self,
+        interval: float = 0.05,  # seconds
+        num_samples: int = 20,
+        directories_to_monitor: Optional[Dict[str, str]] = None,
```
Why not default to `{"main": "/"}`?

AlexandreKempf replied:
I think it is really bad practice to default a dict (or any mutable structure) :)

AlexandreKempf commented on src/dvclive/live.py:
```diff
@@ -370,6 +380,43 @@ def step(self, value: int) -> None:
+
+        Args:
+            interval (float): the time interval between samples in seconds. To keep the
+                sampling interval small, the maximum value allowed is 0.1 seconds.
```
Following W&B behavior on this one.

mattseddon commented:
Should this have closed #81?
